src/libraries/System.Private.CoreLib/src/System/Array.cs

```diff
@@ -785,9 +785,17 @@ public static void Fill<T>(T[] array, T value)
                 ThrowHelper.ThrowArgumentNullException(ExceptionArgument.array);
             }
 
-            for (int i = 0; i < array.Length; i++)
+            if (!typeof(T).IsValueType && array.GetType() != typeof(T[]))
+            {
+                for (int i = 0; i < array.Length; i++)
+                {
+                    array[i] = value;
+                }
+            }
+            else
             {
-                array[i] = value;
+                var span = new Span<T>(ref MemoryMarshal.GetArrayDataReference(array), array.Length);
```

Comment:
Shouldn't need this; should be able to do `new Span<T>(array).Fill(value);`

Reply:
Right, we've already checked `array` for `null` so the null check in `Span(T[]? array)` should be eliminated by the jit. I don't think we can do similar in same `Fill<T>(T[] array, T value, int startIndex, int count)` because the argument validation is more complex.